File: 532-k-diff-pairs-in-an-array/532-k-diff-pairs-in-an-array.py
class Solution:
    def findPairs(self, nums: List[int], k: int) -> int:
        
        prev_num_tracker = {}
        
        ans = set()
        for num in nums:
            
            upper_target_num = k + num
            lower_target_num = num - k
            
            if upper_target_num in prev_num_tracker:
                ans.add((min(upper_target_num, num), max(upper_target_num, num) ) )
            if lower_target_num in prev_num_tracker:
                ans.add((min(lower_target_num, num), max(lower_target_num, num) ) )
            prev_num_tracker[num] = 1
          
        #print(prev_num_tracker)    
        return len(ans)
        

File: 532-k-diff-pairs-in-an-array/532-k-diff-pairs-in-an-array.java
class Solution {
    public int findPairs(int[] nums, int k) {
        HashMap<Integer, Integer> prev_num_tracker = new HashMap<Integer, Integer>();
        Set<Pair> ans = new HashSet<Pair>();
        
        for (int i = 0; i < nums.length; i++) {
            int lower_target = k + nums[i];
            int upper_target = nums[i] - k;
            
            if (prev_num_tracker.containsKey(lower_target)) {
                if (lower_target > nums[i]) {
                    ans.add(new Pair(nums[i], lower_target));
                } else {
                    ans.add(new Pair(lower_target, nums[i]));
                }
            }
            if (prev_num_tracker.containsKey(upper_target)) {
                if (upper_target > nums[i]) {
                    ans.add(new Pair(nums[i], upper_target));
                } else {
                    ans.add(new Pair(upper_target, nums[i]));
                }
            }
            
            prev_num_tracker.put(nums[i], 1);
        }
        
        return ans.size();
        
    }
}